### PR TITLE
fix: unable to reference immutable data in inflight methods

### LINKED ISSUE #1519: Unable to reference data fields in inflight methods of user-defined resources
### Community Note

> Please vote by adding a 👍 reaction to the issue to help us prioritize.
> If you are interested to work on this issue, please leave a comment.

### I tried this:

```js
bring cloud;

resource Foo {
  s: str;
  init() { 
    this.s = "hello";
  }
  inflight test() {
    assert(this.s == "hello");
  }
}

let f = new Foo();
new cloud.Function(inflight () => { f.test(); }) as "test";
```

### I expected this:

Test pass

### Instead, this happened:

```
Error: assertion failed: '(this.s === "hello")'
```

### Component:

Compiler

### Wing Version:

0.5.28

### Node.js Version:

_No response_

### Platform(s):

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by eladb**
Include primitives and collections when identifying preflight fields referenced by inflight methods. Then, for each field, instead of calling `._toInflight()` directly, call a new protected internal method on `Resource` called `._lift(x)` which reuses `serializeImmutableData` to lift the object into an inflight context. Yes, lift...

The test `user-defined-resources-captures` verifies that all types of capturable types can be referenced from inflight methods.

Fixes #1519

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: libs/wingc/src/type_check.rs [hunk 1]
```diff
@@ -515,6 +515,21 @@ impl TypeRef {
...
+			Type::Number => true,
+			Type::String => true,
+			Type::Duration => true,
+			Type::Boolean => true,
+			Type::Array(v) => v.is_capturable(),
+			Type::Map(v) => v.is_capturable(),
+			Type::Set(v) => v.is_capturable(),
+			_ => false,
+		}
+	}
```
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:532]: 👨‍🍳 👌

File: libs/wingc/src/jsify.rs [hunk 2]
```diff
@@ -718,9 +718,9 @@ impl<'a> JSifier<'a> {
 	/// 	}
 	/// }
 	/// _toInflight() {
-	/// 	const b_client = this.b._toInflight();
+	/// 	const b_client = this._lift(b);
```
Comment (Chriscbr) [inline libs/wingc/src/jsify.rs:721]: I wonder if we'll at some point want `lift` to be some kind of method on the inflight host (not the resource). For example, a `cloud.Function` would be responsible for "lifting" a value into a JS string, and a `cloud.Workflow` could "lift" a value into some state machine language value or AST representation, etc.

(for future work)

**Comment by flyingImer** [comment]
this is 🚀, just hit this rough edge and saw this pr!

**Review by Chriscbr — APPROVED** [review]
Pretty elegant approach 👍

File: examples/tests/valid/user-defined-resources-captures.w [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+resource MyResource {
+  my_resource: cloud.Bucket;
+  my_str: str;
+  my_num: num;
+  my_bool: bool;
+  array_of_str: Array<str>;
+  map_of_num: Map<num>;
+  set_of_str: Set<str>;
+
+  array_of_queues: Array<cloud.Queue>;
```
Comment (yoav-steinberg) [inline examples/tests/valid/user-defined-resources-captures.w:12]: Do we also want to add a test for `Array<UserDefinedResource>`?
Also what about:
```wing
struct MyStruct {
  b: cloud.Bucket;
  n: num;
}
```
And testing `Array<MyStruct>`?

File: libs/wingc/src/jsify.rs (hunk 2 above)
Comment (yoav-steinberg) [reply libs/wingc/src/jsify.rs:721]: Interesting concept, but I think as long as out inflights are in a functions/imperitive language (and not a state machine) it makes sense the way it is now.

File: libs/wingc/src/type_check.rs [hunk 4]
```diff
@@ -515,6 +515,21 @@ impl TypeRef {
...
+		match **self {
+			Type::Resource(_) => true,
+			Type::Enum(_) => true,
+			Type::Number => true,
+			Type::String => true,
+			Type::Duration => true,
+			Type::Boolean => true,
+			Type::Array(v) => v.is_capturable(),
+			Type::Map(v) => v.is_capturable(),
+			Type::Set(v) => v.is_capturable(),
```
Comment (yoav-steinberg) [inline libs/wingc/src/type_check.rs:529]: `Type::Struct`?

**Review by yoav-steinberg — APPROVED** [review]
Amazing!

File: libs/wingsdk/src/core/resource.ts [hunk 5]
```diff
@@ -203,6 +204,19 @@ export abstract class Resource
...
+  /**
+   * "Lifts" a value into an inflight context. If the value is a resource (i.e. has a `_toInflight`
+   * method), this method will be called and the result will be returned. Otherwise, the value is
+   * returned as-is.
+   *
+   * @param value The value to lift.
+   * @returns a string representation of the value in an inflight context.
+   * @internal
+   */
+  protected _lift(value: any): string {
```
Comment (yoav-steinberg) [inline libs/wingsdk/src/core/resource.ts:217]: 🛫🛫🛫

**Review by staycoolcall911 — COMMENTED** [review]
Looks like some of the hangar snapshot tests are failing.
You should probably `npm i` (from root folder) before running hangar to update the snapshots (`npx nx hangar test`).

File: libs/wingc/src/jsify.rs (hunk 2 above)
Comment (eladb) [reply libs/wingc/src/jsify.rs:727]: yes, we'll get there! For now, let's say that the resource "lifts" the value into an inflight context.

File: libs/wingc/src/type_check.rs (hunk 4 above)
Comment (eladb) [reply libs/wingc/src/type_check.rs:535]: I can add that but can't really test it yet I think

Comment (eladb) [reply libs/wingc/src/type_check.rs:535]: added

File: examples/tests/valid/user-defined-resources-captures.w (hunk 3 above)
Comment (yoav-steinberg) [reply examples/tests/valid/user-defined-resources-captures.w:12]: We do however have struct literals, I think this should work:
```
array_of_structs: Array<cloud.QueueProps> [
  cloud.QueueProps { timeout: 60s, initialMessages: ["Chicks", "with", "bricks", "come"] },
  cloud.QueueProps { timeout: 120s, initialMessages: ["Chicks", "with", "blocks", "come"] }
];

**Pull request merged**